File: aoc-dec05/dec05.jl
function rowfromstring(s)
    parse(Int, replace(replace(s,"B"=>"1"),"F"=>"0");base=2)
end

function colfromstring(s)
    parse(Int, replace(replace(s,"R"=>"1"),"L"=>"0");base=2)
end

function seat(s)
    r = rowfromstring(s[1:7])
    c = colfromstring(s[8:10]) 
    s = r * 8 + c
    return (row=r,col=c,seatid=s)
end
rowfromstring("BBB")
colfromstring("RLL")


# BFFFBBFRRR: row 70, column 7, seat ID 567.
# FFFBBBFRRR: row 14, column 7, seat ID 119.
# BBFFBBFRLL: row 102, column 4, seat ID 820.

spec_1 = ["BFFFBBFRRR","FFFBBBFRRR","BBFFBBFRLL"]
using Test

seat(spec_1[1])

@test seat(spec_1[1]) == (row = 70,col = 7, seatid = 567)
@test seat(spec_1[2]) == (row=14, col=7, seatid=119)
@test seat(spec_1[3]) == (row=102,col=4,seatid=820)

function seatid(s)
    seat(s).seatid
end

maximum(seatid.(spec_1))
testdata = read(joinpath(@__DIR__,"input.txt"),String)

function dec05_1(s)
    maximum(seatid.(split(s)))
end

@test dec05_1(testdata) == 842

function dec05_2(s)
    list = sort(seatid.(split(s)))
    for (idx,v) in enumerate(list[2:end-1])
        if (list[idx] != v - 1 ) || (list[idx+2] != v + 1)
            @show list[idx] v list[idx+2]
            if list[idx] != v-1
                return v
            else
                return v + 1
            end
        end
    end
end

@test dec05_2(testdata) == 617

File: aoc-dec05/dec05.clj
(require '[clojure.string :as str])

(defn row-from-string [st]
  (Integer/parseInt (str/replace (str/replace st "B" "1") "F" "0") 2))

(defn col-from-string [st]
  (Integer/parseInt (str/replace (str/replace st "R" "1") "L" "0") 2))

(defn seat [st]
  (def r (row-from-string (subs st 0 7)))
  (def c (col-from-string (subs st 7 10)))
  (+ (* r 8) c))

(row-from-string "BBB")
(col-from-string "RLL")

(Integer/parseInt (str/replace (str/replace "BBFBFBF" "B" "0") "F" "1") 2)

(require '[clojure.test :as tst])

(def t1 "BFFFBBFRRR")
(def t2 "FFFBBBFRRR")
(def t3 "BBFFBBFRLL")
(tst/is (= (seat t1) 567))
(tst/is (= (seat t2) 119))
(tst/is (= (seat t3) 820))

(defn dec05-1 [st]
  (apply max (map seat (str/split-lines st))))

(defn dec05-2 [st]
  (compare-next (sort (map seat (str/split-lines st)))))

(defn compare-next [li]
  (if (< (apply - (take 2 li)) -1) ;; Difference > 1 means match
    (inc (first li)) ;; Matching number is missing (between first two in li) 
    (compare-next (next li)))) ;; Else continue searching

(def inputdata (slurp "input.txt"))

(time (tst/is (= (dec05-1 inputdata) 842)))
(time (tst/is (= (dec05-2 inputdata) 617)))